libethdrivers/src/plat/imx6/imx6.c

```diff
@@ -590,11 +591,23 @@ static int init_device(imx6_eth_driver_t *dev)
 
     /* Connect the phy to the ethernet controller */
     unsigned int phy_mask = 0xffffffff;
-    if (fec_init(phy_mask, dev->enet)) {
+    dev->phy = fec_init(phy_mask, dev->enet);
+    if (!dev->phy) {
         ZF_LOGE("Failed to initialize fec");
         goto error;
     }
 
+    enet_set_speed(
+        dev->enet,
+        dev->phy->speed,
+        (dev->phy->duplex == DUPLEX_FULL) ? 1 : 0);
+
+    ZF_LOGI("Link speed: %d Mbps, %s-duplex",
+            dev->phy->speed,
+            (dev->phy->duplex == DUPLEX_FULL) ? "full" : "half");
+
+    udelay(100000); /* why? */
```

Comment:
My guess is that the delay is probably to give time for the link speed setting to propagate.  There's probably a register to poll for completion of the setting, but I wouldn't know what it would be.

Reply:
> My guess is that the delay is probably to give time for the link speed setting to propagate. There's probably a register to poll for completion of the setting, but I wouldn't know what it would be.

Seems the delay is not needed, https://github.com/seL4/util_libs/pull/83/commits/d4cfc2412e7bf91a9056cc869bc09377a2477cf5 removes it completely eventually.

Reply:
Yea I saw this. Because U-boot's model for using ethernet only initializes the hardware directly before use, they probably would see issues with scheduling transmit operations directly after setup without this delay.  Because there's more likely a natural delay from the rest of system setup in the way that the driver's typically used on seL4 this is less of an issue. If you never observe any issue then I'm not going to worry about it.